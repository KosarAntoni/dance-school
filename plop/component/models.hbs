export type {{properCase name}}Props {

}