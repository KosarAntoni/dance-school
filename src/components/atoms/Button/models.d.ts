import { NavBaseProps } from 'atoms/NavBase';
import { TypographyProps } from 'atoms/Typography';

export interface ButtonProps extends NavBaseProps, TypographyProps {
  variant: 'outline' | 'solid' | 'simple' | string;
  space?: 'around' | 'between' | string;
  color?: 'white' | 'blue' | 'black' | 'transparent' | string;
  disabled?: boolean;
}
