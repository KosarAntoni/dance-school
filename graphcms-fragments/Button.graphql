fragment Button on Button {
  stage
  locale
  remoteId: id
  createdAt(variation: COMBINED)
  updatedAt(variation: COMBINED)
  publishedAt(variation: COMBINED)
  label
  slug
  url
  createdBy {
    ... on User {
      remoteTypeName: __typename
      remoteId: id
      stage
    }
  }
  updatedBy {
    ... on User {
      remoteTypeName: __typename
      remoteId: id
      stage
    }
  }
  publishedBy {
    ... on User {
      remoteTypeName: __typename
      remoteId: id
      stage
    }
  }
  breakpoints {
    ... on Breakpoint {
      remoteTypeName: __typename
      remoteId: id
      locale
      stage
    }
  }
  theme
  scheduledIn {
    ... on ScheduledOperation {
      remoteTypeName: __typename
      remoteId: id
      stage
    }
  }
}