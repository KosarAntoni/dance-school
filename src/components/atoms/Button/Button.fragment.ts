import { graphql } from 'gatsby';

export const ButtonFragment = graphql`
  fragment Button on GraphCMS_Button {
    id
    variant
    url
    label
    color
    ariaLabel
  }
`;
