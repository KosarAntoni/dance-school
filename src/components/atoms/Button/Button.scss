@use 'sass:color';
@import 'styles/component-set';

%outline {
  transition: opacity $transition, color $transition, background-color $transition;

  &::after {
    color: inherit;
    position: absolute;
    display: inherit;
    width: calc(100%);
    height: calc(100%);
    border-radius: inherit;
    content: '';
  }
}

%solid {
  position: relative;

  &::after {
    position: absolute;
    display: inherit;
    width: calc(100% + 4px);
    height: calc(100% + 4px);
    border-radius: inherit;
    content: '';
    opacity: 0;
    transition: opacity $transition;
  }

  &:focus,
  &:active {
    &::after {
      opacity: 1;
    }
  }
}

.button {
  position: relative;
  flex-direction: row;
  padding: 0;
  cursor: pointer;
  background: transparent;
  border: none;
  border-radius: 8px;
  transition: background-color $transition;
  @include flex-center;

  &__content {
    display: flex;
    align-items: center;
    width: 100%;
    height: 100%;
  }

  &--solid {
    &-white {
      @extend %solid;

      color: $c-black;
      background-color: $c-white;

      &:hover,
      &:focus {
        background-color: color.adjust($c-white, $lightness: -5%);
      }

      &::after {
        box-shadow: color.adjust($c-white, $lightness: -5%) 0 0 0 2px;
      }
    }

    &-black {
      @extend %solid;

      color: $c-white;
      background-color: $c-black;

      &:hover,
      &:focus {
        background-color: color.adjust($c-black, $lightness: 25%);
      }

      &::after {
        box-shadow: color.adjust($c-black, $lightness: 25%) 0 0 0 2px;
      }
    }

    &-blue {
      @extend %solid;

      color: $c-white;
      background-color: $c-pastel-blue;

      &:hover,
      &:focus {
        background-color: color.adjust($c-pastel-blue, $lightness: 5%);
      }

      &::after {
        box-shadow: color.adjust($c-pastel-blue, $lightness: 5%) 0 0 0 2px;
      }
    }
  }

  &--outline {
    &-white {
      @extend %outline;

      color: $c-black;
      background-color: $c-white;

      &::after {
        box-shadow: $c-black 0 0 0 2px inset;
      }

      &:hover,
      &:focus {
        color: $c-white;
        background-color: $c-black;
      }
    }

    &-black {
      @extend %outline;

      color: $c-white;
      background-color: $c-black;

      &::after {
        box-shadow: $c-white 0 0 0 2px inset;
      }

      &:hover,
      &:focus {
        color: $c-black;
        background-color: $c-white;
      }
    }

    &-blue {
      @extend %outline;

      color: $c-white;
      background-color: $c-pastel-blue;

      &::after {
        box-shadow: $c-white 0 0 0 2px inset;
      }

      &:hover,
      &:focus {
        color: $c-pastel-blue;
        background-color: $c-white;
      }
    }
  }

  &--simple {
    &-white {
      color: $c-white;

      &:hover,
      &:focus {
        color: color.adjust($c-white, $lightness: -5%);
      }
    }

    &-black {
      color: $c-black;

      &:hover,
      &:focus {
        color: color.adjust($c-black, $lightness: 5%);
      }
    }

    &-blue {
      color: $c-pastel-blue;

      &:hover,
      &:focus {
        color: color.adjust($c-pastel-blue, $lightness: 5%);
      }
    }
  }

  &--around {
    .button {
      &__content {
        justify-content: center;
      }
    }
  }

  &--between {
    .button {
      &__content {
        justify-content: space-between;
      }
    }
  }

  &--disabled {
    opacity: 0.5;
  }
}
