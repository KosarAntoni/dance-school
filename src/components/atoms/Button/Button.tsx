import React, { FC } from 'react';
import classNames from 'classnames';

import NavBase from 'atoms/NavBase';
import Typography from 'atoms/Typography';

import { ButtonProps } from './models.d';

import './Button.scss';

const Button: FC<ButtonProps> = ({
  children,
  variant = 'solid',
  space = 'around',
  color = 'blue',
  disabled,
  className,
  onClick,
  url,
  ariaLabel,
  lineHeight,
  weight,
  italic,
  size,
  padding,
}) => {
  const buttonClasses = classNames('button', className, {
    [`button--${variant}-${color}`]: variant && color,
    [`button--${space}`]: space,
    'button--disabled': disabled,
  });

  const buttonContentClasses = classNames('button__content');

  const navBaseProps = { onClick, url, ariaLabel };
  const typographyProps = {
    lineHeight,
    weight,
    italic,
    size,
    padding,
  };

  return (
    <NavBase className={buttonClasses} {...navBaseProps}>
      <Typography as="span" {...typographyProps} className={buttonContentClasses}>
        {children}
      </Typography>
    </NavBase>
  );
};

export default Button;
