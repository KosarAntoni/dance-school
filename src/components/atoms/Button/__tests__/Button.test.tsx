import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import {
  ButtonChildrenMock,
  buttonDefaultClassesMock,
  ButtonDefaultPropsMock,
  testFunction,
} from '../__mock__/mock';
import Button from '../Button';

import { ButtonProps } from '../models';

const defaultProps: ButtonProps = {
  ...ButtonDefaultPropsMock,
};

test('the Button component renders correctly with required and optional props', () => {
  render(<Button {...defaultProps}>{ButtonChildrenMock}</Button>);

  const buttonNode = screen.getByRole('button');

  // the component has been rendered
  expect(buttonNode).toBeInTheDocument();

  // generating a snapshot, not for testing purposes but only to see what changes in the HTML structure during CR if the developer updates the snapshot
  expect(buttonNode).toMatchSnapshot();

  // after click right function is called
  fireEvent.click(buttonNode);
  expect(testFunction).toBeCalled();

  // the comonent has correct classes
  expect(buttonNode).toHaveClass(buttonDefaultClassesMock.join(' '));
});
