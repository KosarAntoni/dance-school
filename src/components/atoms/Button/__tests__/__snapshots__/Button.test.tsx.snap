// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Button component renders correctly with required and optional props 1`] = `
<button
  class="button test-class button--solid-blue button--around"
  type="button"
>
  <span
    class="typography button__content typography--size-20"
    data-testid="typography-item"
  >
    test content
  </span>
</button>
`;
